From: Egbert Eich <eich@suse.de>
Date: Thu Aug 28 19:51:50 2014 +0200
Subject: drm/panic: Only restore console if kernel isn't set to reboot on panic
Patch-mainline: to be upstreamed

References: bnc#892114
Signed-off-by: Egbert Eich <eich@suse.com>

The mode restoration in fb_helper_panic() is potentially dangerous
as the mode setting code called from here was never meant to be called
from a panic context.
However another panic in this code may cause the system to hang and
prevent the reboot.
Therefore only restore console when it is set to wait for ever (which
is the default).

Signed-off-by: Egbert Eich <eich@suse.de>
---
 drivers/gpu/drm/drm_fb_helper.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/drm_fb_helper.c b/drivers/gpu/drm/drm_fb_helper.c
index 49557c9..92d077e 100644
--- a/drivers/gpu/drm/drm_fb_helper.c
+++ b/drivers/gpu/drm/drm_fb_helper.c
@@ -344,9 +344,10 @@ static int drm_fb_helper_panic(struct notifier_block *n, unsigned long ununsed,
 {
 	/*
 	 * It's a waste of time and effort to switch back to text console
-	 * if the kernel should reboot before panic messages can be seen.
+	 * if the kernel should reboot - since the fb_panic code may hang
+	 * it may even prevent rebooting.
 	 */
-	if (panic_timeout < 0)
+	if (panic_timeout != 0)
 		return 0;
 
 	pr_err("panic occurred, switching back to text console\n");
